### PR TITLE
Propagate builtins pass

### PR DESCRIPTION
**Pull request by tehrengruber**
Propagate calls to deref into lambda functions.

Transform:
```
·(λ(inner_it) → lift(stencil)(inner_it))(outer_it)
```
into:
```
(λ(inner_it) → ·(lift(stencil)(inner_it)))(outer_it)
```
After this pass calls to deref are grouped together (better) with lambda functions being propagated outwards. This increases the ability of the lift inliner to remove the deref + lift combination.

### CODE REVIEW

**Review by havogt — CHANGES_REQUESTED** [review]
Did you think about other builtins?
- Is it obvious that it works on every builtin? To me it wasn't, so I checked for `lift` and it seems to make sense.
- It does it only for unary builtins, which limits already to very few.
Therefore, maybe we should limit to `deref` or the ones that we know make sense.

File: src/functional/iterator/transforms/propagate_builtins.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        after which the lift inliner can remove the deref + lift combination.
+        """
+        return cls().visit(node)
+
+    def visit_FunCall(self, node: ir.FunCall):
+        if P(ir.FunCall, fun=P(ir.SymRef), args=[P(ir.FunCall, fun=P(ir.Lambda))]).match(node):
+            builtin = node.fun
+            lambda_fun: ir.Lambda = node.args[0].fun  # type: ignore[attr-defined]
+            lambda_args: list[ir.Expr] = node.args[0].args  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/iterator/transforms/propagate_builtins.py]: I agree, we should introduce a convention and add them to the guidelines. I added a comment here.

File: src/functional/iterator/transforms/propagate_builtins.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        are propagated inwards::
+
+            (λ(inner_it) → ·(lift(stencil)(inner_it)))(outer_it)
+
+        after which the lift inliner can remove the deref + lift combination.
+        """
+        return cls().visit(node)
+
+    def visit_FunCall(self, node: ir.FunCall):
+        if P(ir.FunCall, fun=P(ir.SymRef), args=[P(ir.FunCall, fun=P(ir.Lambda))]).match(node):
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/functional/iterator/transforms/propagate_builtins.py]: Indeed, I generalized this pass to more builtins, but wasn't thorough enough and overlooked binary builtins (which are more work due to potential symbol collisions). I reverted to only handle `deref` and added a todo to generalize at a later time.

File: src/functional/iterator/transforms/propagate_deref.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+from eve import NodeTranslator
+from eve.pattern_matching import ObjectPattern as P
+from functional.iterator import ir
+
+
+# TODO(tehrengruber): This pass can be generalized to all builtins, e.g.
+#  `plus((λ(...) → plus(...))(...), ...)` can be transformed into
```
Comment (havogt) [inline src/functional/iterator/transforms/propagate_deref.py:7]: isn't the inner plus confusing, i.e. unnecessary for what you describe?

**Review by havogt — APPROVED** [review]
Looks good.

File: src/functional/iterator/transforms/propagate_deref.py (hunk 3 above)
Comment (tehrengruber) [reply src/functional/iterator/transforms/propagate_deref.py]: True. Fixed.

**Pull request merged**